nav2_map_server/src/map_saver.cpp

```diff
@@ -0,0 +1,215 @@
+      for(unsigned int y = 0; y < map->info.height; y++) {
+        for(unsigned int x = 0; x < map->info.width; x++) {
+          unsigned int i = x + (map->info.height - y - 1) * map->info.width;
+          if (map->data[i] >= 0 && map->data[i] <= threshold_free_) {  // [0,free)
+            fputc(254, out);
+          } else if (map->data[i] >= threshold_occupied_) {  // (occ,255]
+            fputc(000, out);
+          } else {  // occ [0.25,0.65]
+            fputc(205, out);
+          }
+        }
+      }
+
+      fclose(out);
+
+
+      std::string mapmetadatafile = mapname_ + ".yaml";
+      RCLCPP_INFO(logger, "Writing map occupancy data to %s", mapmetadatafile.c_str());
+      FILE* yaml = fopen(mapmetadatafile.c_str(), "w");
+
+
+      /*
+resolution: 0.100000
+origin: [0.000000, 0.000000, 0.000000]
+#
+negate: 0
+occupied_thresh: 0.65
+free_thresh: 0.196
+
+       */
+
+      geometry_msgs::msg::Quaternion orientation = map->info.origin.orientation;
+      tf2::Matrix3x3 mat(tf2::Quaternion(
+        orientation.x,
+        orientation.y,
+        orientation.z,
+        orientation.w
+      ));
+      double yaw, pitch, roll;
+      mat.getEulerYPR(yaw, pitch, roll);
+
+      fprintf(yaml, "image: %s\nresolution: %f\norigin: [%f, %f, %f]\n",
+              mapdatafile.c_str(), map->info.resolution,
+              map->info.origin.position.x, map->info.origin.position.y, yaw);
+      fprintf(yaml, "negate: 0\noccupied_thresh: 0.65\nfree_thresh: 0.196\n\n");
+
+      fclose(yaml);
+
+      RCLCPP_INFO(logger, "Done\n");
+      saved_map_ = true;
+    }
+
+    std::string mapname_;
+    rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::ConstSharedPtr map_sub_;
+    bool saved_map_;
+    int threshold_occupied_;
+    int threshold_free_;
+};
+
+#define USAGE "Usage: \n" \
+              "  map_saver -h\n"\
+              "  map_saver [--occ <threshold_occupied>] [--free <threshold_free>] " \
+              "[-f <mapname>] [ROS remapping args]"
+
+
+int main(int argc, char** argv)
+{
+  rclcpp::init(argc, argv);
+  rclcpp::Logger logger = rclcpp::get_logger("map_saver");
+
+  std::string mapname = "map";
+  int threshold_occupied = 65;
+  int threshold_free = 25;
+
+  for(int i = 1; i < argc; i++)
+  {
+    if(!strcmp(argv[i], "-h"))
+    {
+      puts(USAGE);
+      return 0;
+    } else if(!strcmp(argv[i], "-f")) {
+      if(++i < argc)
+        mapname = argv[i];
+      else
+      {
+        puts(USAGE);
+        return 1;
+      }
+    } else if (!strcmp(argv[i], "--occ")) {
+      if (++i < argc)
+      {
+        threshold_occupied = atoi(argv[i]);
+        if (threshold_occupied < 1 || threshold_occupied > 100)
+        {
+          RCLCPP_ERROR(logger, "Threshold_occupied must be between 1 and 100");
+          return 1;
+        }
+      } else {
+        puts(USAGE);
+        return 1;
+      }
+    } else if (!strcmp(argv[i], "--free")) {
+      if (++i < argc)
+      {
+        threshold_free = atoi(argv[i]);
```

Comment:
Since this is a straight port, I didn't change any includes but I'll add any that are missing

Reply:
That's fine. I just saw that you removed the std:: from atoi as part of the port, which seems like it should make it not compile. Since it does, there must be namespace pollution upstream.

Just pointing it out.